### PR TITLE
Allow autoloaders to require/require_once for #4836 via a fake read

### PR DESCRIPTION
**Pull request by rrazor**
## Overview

[PHPStan issue #4836](https://github.com/phpstan/phpstan/issues/4836) occurs when a custom autoloader uses `require` or `require_once` and PHPStan tries to process a file that contains an anonymous class definition with dependencies that must be autoloaded (see issue for details).

## Approach

The goal of `FileReadTrapStreamWrapper` is to use a custom autoloader's knowledge of where to find a file, grab the path, and shut down the rest of the `include` by returning `false` from `stream_open`.

This PR allows `require` and `require_once` to be used in a custom autoloader by allowing the entire file read to occur, with an empty string returned from `stream_read()` to serve as dummy data that is also valid PHP.

## Testing

I ran the complete PHPStan test suite and observed no errors. I also confirmed that the updated code succeeded in my [sample repository](https://github.com/rrazor/phpstan-autoload-error-example).

## Potential issues

Since we simulate a file read from beginning to end, `include_once` and `require_once` will not try to load the same path again. **I think this is okay because I assume PHPStan loads the file itself later.**

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
I copied your test project to PHPStan's E2E tests: https://github.com/phpstan/phpstan/commit/2bef7e4e48d1ab5bf47558a58543810ad9ad7057

After it fails the build, I'll merge this. Thank you very much!

**Comment by ondrejmirtes** [comment]
BTW Feel free to also submit this to https://github.com/Roave/BetterReflection, thank you.

File: src/Reflection/BetterReflection/SourceLocator/FileReadTrapStreamWrapper.php [hunk 1]
```diff
@@ -86,7 +86,51 @@ public function stream_open($path, $mode, $options, &$openedPath): bool
...
+	 * @param int $count
+	 *
+	 * @return string
+	 */
+	public function stream_read($count): string
+	{
+		// Dummy return value that is also valid PHP for require(). We'll read
+		// and process the file elsewhere, so it's OK to provide dummy data for
+		// this read.
+		return '';
```
Comment (dktapps) [inline src/Reflection/BetterReflection/SourceLocator/FileReadTrapStreamWrapper.php:105]: This causes problems when OPcache is enabled. OPcache will remember this empty string when the file is next autoloaded (which might not be within FileReadTrapStreamWrapper).

Comment (dktapps) [reply src/Reflection/BetterReflection/SourceLocator/FileReadTrapStreamWrapper.php:105]: It also causes problems when `require_once` or `include_once` is used, regardless of whether opcache is enabled or not.

**Pull request merged**